### ISSUE TITLE
Eslint Does Not Detect Custom Created Plugin Rule

### ISSUE THREAD

**Issue by turturmurmur** [eslint/eslint#16859]
### Environment

Node version: v14.18.1
npm version:  v6.14.15
Local ESLint version: v8.33.0 (Currently used)
Global ESLint version: Not found
Operating System: win32 10.0.19045

### What parser are you using?

@typescript-eslint/parser

### What did you do?

Created a custom plugin in my local where I wanted that to check .json files for any duplicate values. Did install the package locally and added the rule to the eslintrc.json file.
Any time someone tries to add a duplicate value in a json file(not a duplicate key) I wanted that plugin to return a rule error.
I have used, eslint, typescript, prettier, and auto format and fix on save. using vscode.
Attached a zip file of the demo project to this thread below.
[prettier,eslint,typescript setup.zip](https://github.com/eslint/eslint/files/10610402/prettier.eslint.typescript.setup.zip)

### What did you expect to happen?

Find the duplicate values in the json files(not the keys)

### What actually happened?

Did not find any errors when I manually added 2 different keys with 2 duplicate values. But other typescript rules worked and were found.
![image](https://user-images.githubusercontent.com/89113207/216806388-b7a669ba-00cf-4912-9ac5-020360f013a8.png)

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by yeonjuan** [comment]
Hi @turturmurmur It's because of the `fs.readdir`. The `context.report()` may be called after eslint aggregates reports. Try to use `fs.readdirSync`.

```js
// Don't
fs.readdir(folder, (err, files) => {
            //...
      });
// Try
fs.readdirSync(...);
```

It's not a bug so I close this issue

**Comment by turturmurmur** [comment]
Hi @yeonjuan , I have tried with that option too but it still did not work.
![image](https://user-images.githubusercontent.com/89113207/216830601-8083bf85-18e0-4cb4-9c8d-0e60355f95fd.png)
Then I tried re-writing the function hopelessly as below too, but no luck.
![image](https://user-images.githubusercontent.com/89113207/216830774-5e094d9a-8f36-485a-80ad-c140da93ae20.png)

Can you take a look at it, and tell me what might be wrong, please?

**Comment by yeonjuan** [comment]
@turturmurmur I do not know well about your code and the libraries you use. So I'd like to suggest you try simple code to checks whether it's an eslint issue or not.

When i tested with simplified code like the one below, it worked! So I think there is a bug in your code.

```js
module.exports = {
  rules: {
    'no-duplicate-values': {
      meta: {
        type: 'problem',
        docs: {
          description: 'Disallow duplicate values in JSON files',
          category: 'Possible Errors',
          recommended: 'error',
          url: 'https://eslint.org/docs/rules/no-duplicate-values',
        },
        schema: [],
      },
      create: function (context) {
        return {
          Program: function (node) {
            context.report({
              node: node,
              message: `Duplicate value found in file`,
            });
          },
        };
      },
    },
  },
};

```

- output

```console
1:1  error    Duplicate value found in file              no-duplicate-values/no-duplicate-values
  5:7  warning  'book' is assigned a value but never used  @typescript-eslint/no-unused-vars
```

**Comment by turturmurmur** [comment]
@yeonjuan , you were correct. I removed the dependency and re-wrote the code. Now it finds the errors. Thank you for that tip.
However, after that one weird issue came out. Eslint thinks the errors are in a different file.
![image](https://user-images.githubusercontent.com/89113207/216841206-42f216bc-c5aa-47ad-ae1f-2067971ac316.png)

To reproduce I am sending a zip of the project below. 
[prettier,eslint,typescript setup.zip](https://github.com/eslint/eslint/files/10611996/prettier.eslint.typescript.setup.zip)

I don't know how I can make sure it sets the resource path correctly. Can you help with this issue, please?

**Comment by Rec0iL99** [comment]
@turturmurmur 

> @yeonjuan , you were correct. I removed the dependency and re-wrote the code. Now it finds the errors. Thank you for that tip. However, after that one weird issue came out. Eslint thinks the errors are in a different file. ![image](https://user-images.githubusercontent.com/89113207/216841206-42f216bc-c5aa-47ad-ae1f-2067971ac316.png)
> 
> To reproduce I am sending a zip of the project below. [prettier,eslint,typescript setup.zip](https://github.com/eslint/eslint/files/10611996/prettier.eslint.typescript.setup.zip)
> 
> I don't know how I can make sure it sets the resource path correctly. Can you help with this issue, please?

@turturmurmur has opened a separate issue for this at #16867
